File: solutions/126.Word_Ladder/AC_two_end_BFS_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_two_end_BFS_n.py
# Create Date: 2015-03-29 15:10:46
# Usage:       AC_two_end_BFS_n.py 
# Descripton:  


import string
class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return an integer
    def ladderLength(self, start, end, dict):
        start_set, end_set = set([start]), set([end])

        n = len(start)
        dis = 1
        chars = string.ascii_lowercase

        while start_set and end_set:
            dis += 1
            if len(start_set) < len(end_set):
                start_set, end_set = end_set, start_set
            new_start_set = set()
            for word in start_set:
                for idx in range(n):
                    for c in chars:
                        s = word[:idx] + c + word[idx+1:]
                        if s in end_set:
                            return dis
                        if s in dict:
                            new_start_set.add(s)
                            dict.remove(s)
            start_set = new_start_set
            
        return 0


# debug
s = Solution()
print s.ladderLength("hot", "dog", ["hot","dog"])

File: solutions/126.Word_Ladder/AC_two_end_BFS_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_two_end_BFS_n.cpp
*  Create Date: 2015-03-29 14:11:47
*  Descripton:   
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
// it cost a lot of time
// private:
//     bool checkChar(string &str1, string &str2) {
//         int diff = 0;
//         for (int i = 0; i < str1.length(); ++i) {
//             if (str1[i] != str2[i])
//                 ++diff;
//             if (diff > 1)
//                 return false;
//         }
//         return diff == 1;
//     }

public:
    int ladderLength(string start, string end, unordered_set<string> &dict) {
        // use unordered_set as queue
        unordered_set<string> startSet, endSet, *set1, *set2;

        startSet.insert(start);
        endSet.insert(end);
        int dis = 1, len = start.length();
        while (!startSet.empty() && !endSet.empty()) {
            if (startSet.size() > endSet.size()) {
                set1 = &startSet;
                set2 = &endSet;
            } else {
                set1 = &endSet;
                set2 = &startSet;
            }

            ++dis;

            unordered_set<string> newset1;
            for (auto itr = set1->begin(); itr != set1->end(); itr++) {
                string old = *itr;
                for (int i = 0; i < len; ++i) {
                    char tmp = old[i];
                    for (int j = 0; j < 26; ++j) {
                        old[i] = 'a' + j;
                        auto f = set2->find(old);
                        if (f != set2->end())
                            return dis;
                        f = dict.find(old);
                        if (f != dict.end()) {
                            newset1.insert(old);
                            dict.erase(f);
                        }
                    }
                    old[i] = tmp;
                }
            }
            swap(newset1, *set1);
        }
        return 0;
    }
};

int main() {

    return 0;
}

